lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -350,3 +353,201 @@ def _get(self, json_data: JSON, _) -> Response:
         """
         return self._call(requests.get, None)
+
+
+class ElasticBulkBase(ApiBase):
+    """
+    A base class for bulk Elasticsearch queries that allows subclasses to
+    provide a generator to produce bulk command documents with common setup and
+    results processing.
+
+    This class extends the ApiBase class in order to connect the post
+    and get methods to Flask's URI routing algorithms. It implements a common
+    mechanism for calling Elasticsearch and processing errors.
+    """
+
+    def __init__(
+        self,
+        config: PbenchServerConfig,
+        logger: Logger,
+        schema: Schema,
+        *,  # following parameters are keyword-only
+        role: API_OPERATION = API_OPERATION.UPDATE,
+    ):
+        """
+        Base class constructor.
+
+        Args:
+            config: server configuration
+            logger: logger object
+            schema: API schema: for example,
+                    Schema(
+                        Parameter("user", ParamType.USER, required=True),
+                        Parameter("start", ParamType.DATE)
+                    )
+            role: specify the API role, defaulting to READ
+        """
+        super().__init__(config, logger, schema, role=role)
+        host = config.get("elasticsearch", "host")
+        port = config.get("elasticsearch", "port")
+        self.es_url = f"http://{host}:{port}"
+
+    def generate_documents(self, json_data: JSON, dataset: Dataset) -> Iterator[dict]:
+        """
+        Generate a series of Elasticsearch bulk commands to be fed to the
+        streaming_bulk helper.
+
+        This is an abstract method that must be implemented by a subclass.
+
+        Args:
+            json_data: JSON dictionary of type-normalized key-value pairs
+                controller: the controller that generated the dataset
+                name: name of the dataset to publish
+                access: The desired access level of the dataset
+
+            dataset: The associated Dataset object
+        """
+        raise NotImplementedError()
+
+    def complete(self, dataset: Dataset, json_data: JSON, error_count: int) -> None:
+        """
+        Complete a bulk Elasticsearch operation, generally by modifying the
+        source Dataset resource.
+
+        This is an abstract method that must be implemented by a subclass.
+
+        Args:
+
+            dataset: The associated Dataset object
+
+            json_data: JSON dictionary of type-normalized key-value pairs
+                controller: the controller that generated the dataset
+                name: name of the dataset to publish
+                access: The desired access level of the dataset
+
+            error_count: The number of errors encountered during the bulk
+                operation. We want to support idempotency: the Dataset should
+                not be altered unless the error_count is 0.
+        """
+        raise NotImplementedError()
```

Comment:
I suppose in this case we could make the abstract superclass behavior a no-op so that a subclass wouldn't need to override it if there's no completion action. (On the other hand, publish does need a completer, and delete's will be vastly more complex, so I'm not sure there's a real use case for that idea.)